File: p108e/sorted_array_to_bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const Tree = require('binary_tree');
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
var sortedArrayToBST = function (nums) {
  if (nums.length == 0) return null;
  let mid = ~~(nums.length / 2);
  let root = new Tree.TreeNode(nums[mid]);
  root.left = sortedArrayToBST(nums.slice(0, mid));
  root.right = sortedArrayToBST(nums.slice(mid + 1));
  return root;
};
// TESTS
[
  [[-10, -3, 0, 5, 9], '0,-3,-10,#,#,#,9,5,#,#,#'],
  [[1, 3], '3,1,#,#,#'],
  [[], '#'],
].forEach(([nums, expected]) => {
  const actual = Tree.serialize(sortedArrayToBST(nums));
  console.log('Convert', nums, 'to BST ->', actual);
  console.assert(actual === expected);
});

File: p108e/sorted_array_to_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from local_packages.binary_tree import TreeNode
from typing import List


class Solution:
    def sortedArrayToBST(self, nums: List[int]) -> TreeNode:
        if not nums:
            return None
        mid = len(nums) // 2
        return TreeNode(
            nums[mid],
            self.sortedArrayToBST(nums[:mid]),
            self.sortedArrayToBST(nums[mid + 1 :]),
        )


# TESTS
for nums, expected in [
    ([-10, -3, 0, 5, 9], "0,-3,-10,#,#,#,9,5,#,#,#"),
    ([1, 3], "3,1,#,#,#"),
    ([], "#"),
]:
    sol = Solution()
    actual = TreeNode.serialize(sol.sortedArrayToBST(nums))
    print("Convert", nums, "to BST ->", actual)
    assert actual == expected
